user/user.controller.js

```diff
@@ -1,29 +1,99 @@
 const { User, validate } = require('./user.model')
+const { Token, resetEmailValidate, confirmResetPassword } = require('./token.model')
+const sendEmail = require('../utils/emails/sendEmail')
 const bcrypt = require('bcrypt')
+const crypto = require('crypto')
 const _ = require('lodash')
 
 exports.register = async (req, res) => {
     const { error } = validate(req.body)
     if (error) return res.status(400).send({ detail: error.details[0].message })
 
-    try {
-        let user = await User.findOne({ email: req.body.email })
-        if (user) return res.status(400).send({'detail': 'User with email already exists.'})
+    let user = await User.findOne({ email: req.body.email })
+    if (user) return res.status(400).send({ 'detail': 'User with email already exists.' })
 
-        requestBody = _.pick(req.body, ['_id', 'firstName', 'lastName', 'email', 'password'])
+    requestBody = _.pick(req.body, ['_id', 'firstName', 'lastName', 'email', 'password'])
 
-        if(requestBody.password !== req.body.confirmPassword) return res.status(400).send({detail: 'Password Mismatch'})
+    if (requestBody.password !== req.body.confirmPassword) return res.status(400).send({ detail: 'Password Mismatch' })
 
-        user = new User(requestBody)
-        const salt = await bcrypt.genSalt(10)
-        user.password = await bcrypt.hash(user.password, salt)
+    user = new User(requestBody)
+    const salt = await bcrypt.genSalt(10)
+    user.password = await bcrypt.hash(user.password, salt)
 
-        await user.save()
-        const data = _.pick(user, ['_id', 'firstName', 'lastName', 'email'])
-        res.send(data)
+    await user.save()
+    const data = _.pick(user, ['_id', 'firstName', 'lastName', 'email'])
+
+    const token = user.generateAuthToken()
+    res.header('x-auth-token', token).send(data)
+}
+
+exports.requestResetPassword = async (req, res) => {
+    const clientURL = 'http://localhost:8000/api/users'
+    const { error } = resetEmailValidate(req.body)
+    if (error) return res.status(400).send({ detail: error.details[0].message })
+
+    const user = await User.findOne({email: req.body.email})
+
+    if (!user) return res.status(400).status('User does not exist')
+
+    let token = await Token.findOne({ userId: user._id })
+    if (token) await token.deleteOne()
+
+    let resetToken = crypto.randomBytes(32).toString('hex')
+    const hash = await bcrypt.hash(resetToken, Number(10))
```

Comment:
Are you sure you need to hash this? Considering you can still see the original token in the URL, does hashing and comparing (an expensive process) improve security?